File: hexameter-core/src/main/java/org/codetome/hexameter/core/api/HexagonalGrid.java
package org.codetome.hexameter.core.api;

import org.codetome.hexameter.core.backport.Optional;
import org.codetome.hexameter.core.internal.GridData;
import rx.Observable;

import java.util.Collection;

/**
 * <p>
 * Represents a hexagonal grid. Use {@link HexagonalGridBuilder} to generate a
 * ready-to-use grid. This interface contains all common functionality for dealing with
 * Hexagons. See {@link HexagonalGridCalculator} for more advanced features.
 * </p>
 * <p>
 * This {@link HexagonalGrid} uses an cube (trapezoidal) coordinate system for easier
 * computation. This means that apart from the X axis a diagonal axis is used instead of
 * the vertical Y axis.
 * </p>
 */
public interface HexagonalGrid {

    /**
     * Returns this HexagonalGrid's GridData.
     *
     * @return grid data
     */
    GridData getGridData();

    /**
     * Returns all {@link Hexagon}s contained in this grid.
     *
     * @return hexagons
     */
    Observable<Hexagon> getHexagons();

    /**
     * Returns all {@link Hexagon}s contained in the given cube coordinate range.
     * If the range contains coordinates which are not part of the grid they will be ignored.
     *
     * @param from from
     * @param to to
     *
     * @return {@link Hexagon}s in the given range.
     */
    Observable<Hexagon> getHexagonsByCubeRange(CubeCoordinate from, CubeCoordinate to);

    /**
     * Returns all {@link Hexagon}s contained in the given offset coordinate range.
     * If the range contains coordinates which are not part of the grid they will be ignored.
     *
     * @param gridXFrom from x inclusive
     * @param gridXTo to x inclusive
     * @param gridYFrom from z inclusive
     * @param gridYTo to z inclusive
     *
     * @return {@link Hexagon}s in the given range.
     */
    Observable<Hexagon> getHexagonsByOffsetRange(int gridXFrom, int gridXTo, int gridYFrom, int gridYTo);

    /**
     * Tells whether the given cube coordinate is on the grid or not.
     * If you want to look up by offset coordinate use {@link CoordinateConverter}.
     *
     * @param coordinate coord
     *
     * @return is it on the grid?
     */
    boolean containsCubeCoordinate(CubeCoordinate coordinate);

    /**
     * Returns a {@link Hexagon} by its cube coordinate.
     *
     * @param coordinate coord
     *
     * @return Optional with a Hexagon if it is present
     */
    Optional<Hexagon> getByCubeCoordinate(CubeCoordinate coordinate);

    /**
     * Returns a {@link Hexagon} by a pixel coordinate.
     * <em>Please note</em> that all pixel coordinates are relative to
     * the containing {@link HexagonalGrid}.
     *
     * @param coordinateX pixel coordinateX coordinate
     * @param coordinateY pixel coordinateY coordinate
     *
     * @return Optional with a Hexagon if it is present
     */
    Optional<Hexagon> getByPixelCoordinate(double coordinateX, double coordinateY);

    /**
     * Returns a neighbor of a Hexagon by its neighbor index.
     *
     * @return neighbor or empty Optional if not applicable
     */
    Optional<Hexagon> getNeighborByIndex(Hexagon hexagon, int index);

    /**
     * Returns all neighbors of a {@link Hexagon}.
     *
     * @param hexagon {@link Hexagon}
     *
     * @return the {@link Hexagon}'s neighbors
     */
    Collection<Hexagon> getNeighborsOf(Hexagon hexagon);

    /**
     * Clears all satellite data attached to the {@link Hexagon}s in this grid.
     */
    void clearSatelliteData();
}

File: mixite.core/src/commonMain/kotlin/org/hexworks/mixite/core/api/HexagonalGrid.kt
package org.hexworks.mixite.core.api

import org.hexworks.cobalt.datatypes.Maybe
import org.hexworks.mixite.core.api.contract.SatelliteData
import org.hexworks.mixite.core.internal.GridData

/**
 *
 *
 * Represents a hexagonal grid. Use [HexagonalGridBuilder] to generate a
 * ready-to-use grid. This interface contains all common functionality for dealing with
 * Hexagons. See [HexagonalGridCalculator] for more advanced features.
 *
 *
 *
 * This [HexagonalGrid] uses an cube (trapezoidal) coordinate system for easier
 * computation. This means that apart from the X axis a diagonal axis is used instead of
 * the vertical Y axis.
 *
 */
interface HexagonalGrid<T : SatelliteData> {

    /**
     * Returns this HexagonalGrid's GridData.
     *
     * @return grid data
     */
    val gridData: GridData

    /**
     * Returns all [Hexagon]s contained in this grid.
     *
     * @return hexagons
     */
    val hexagons: Iterable<Hexagon<T>>

    /**
     * Returns all [Hexagon]s contained in the given cube coordinate range.
     * If the range contains coordinates which are not part of the grid they will be ignored.
     *
     * @param from from
     * @param to to
     *
     * @return [Hexagon]s in the given range.
     */
    fun getHexagonsByCubeRange(from: CubeCoordinate, to: CubeCoordinate): Iterable<Hexagon<T>>

    /**
     * Returns all [Hexagon]s contained in the given offset coordinate range.
     * If the range contains coordinates which are not part of the grid they will be ignored.
     *
     * @param gridXFrom from x inclusive
     * @param gridXTo to x inclusive
     * @param gridYFrom from z inclusive
     * @param gridYTo to z inclusive
     *
     * @return [Hexagon]s in the given range.
     */
    fun getHexagonsByOffsetRange(gridXFrom: Int, gridXTo: Int, gridYFrom: Int, gridYTo: Int): Iterable<Hexagon<T>>

    /**
     * Tells whether the given cube coordinate is on the grid or not.
     * If you want to look up by offset coordinate use [CoordinateConverter].
     *
     * @param coordinate coord
     *
     * @return is it on the grid?
     */
    fun containsCubeCoordinate(coordinate: CubeCoordinate): Boolean

    /**
     * Returns a [Hexagon] by its cube coordinate.
     *
     * @param coordinate coord
     *
     * @return Maybe with a Hexagon if it is present
     */
    fun getByCubeCoordinate(coordinate: CubeCoordinate): Maybe<Hexagon<T>>

    /**
     * Returns a [Hexagon] by a pixel coordinate.
     * *Please note* that all pixel coordinates are relative to
     * the containing [HexagonalGrid].
     *
     * @param coordinateX pixel coordinateX coordinate
     * @param coordinateY pixel coordinateY coordinate
     *
     * @return Maybe with a Hexagon if it is present
     */
    fun getByPixelCoordinate(coordinateX: Double, coordinateY: Double): Maybe<Hexagon<T>>

    /**
     * Returns the coordinate of a neighbor of a Hexagon by its neighbor index.
     *
     * @return CubeCoordinate
     */
    fun getNeighborCoordinateByIndex(coordinate: CubeCoordinate, index: Int): CubeCoordinate

    /**
     * Returns a neighbor of a Hexagon by its neighbor index.
     *
     * @return neighbor or empty Maybe if not applicable
     */
    fun getNeighborByIndex(hexagon: Hexagon<T>, index: Int): Maybe<Hexagon<T>>

    /**
     * Returns all neighbors of a [Hexagon].
     *
     * @param hexagon [Hexagon]
     *
     * @return the [Hexagon]'s neighbors
     */
    fun getNeighborsOf(hexagon: Hexagon<T>): Collection<Hexagon<T>>

}
